### DISCUSSION TITLE
SWAP Partition bug

### DISCUSSION THREAD

**Discussion by asmcc** [Freetz-NG/freetz-ng#725] [Allgemein] [answered]
Hallo fda77 / cuma,

ich habe unter "Pull requests" eine kleine Berichtigung für udev gepostet, die es erlaubt, SWAP-Partitionen auf USB-Medien korrekt zu behandeln. Ich könnte es natürlich auch als Issue melden, denn momentan werden die SWAP-Partitionen von udev als fehlerhaft erkannt, rc.swap startet mit dieser SWAP-Partition dann nicht automatisch und muss nachher manuell angestoßen werden. Das alles wird mit dem Pull Request "geheilt".

Wie im Pull Request beschrieben, habe ich diese Änderung auf meiner 7590 erfolgreich getestet und erwarte davon keine Nebenwirkungen auch auf den anderen Boxen.

Wäre schön, wenn du es einchecken könntest, bevor es durch "stalled" ganz in Vergessenheit gerät.

Danke!

**Comment by fda77** [comment]
Sorry, bin nicht dazu gekommen ein paar Dinge auszuprobieren.
Wie du gesehen hast mountet udevmount nicht selbst, sondern ist ein zwischen Layer. Swap-Partitionen hat AVM nicht. Was mir hier nicht so ganz gefällt dass die ganze Logik im nicht-erfolgreich Block liegt. Du hast dich ja beschwert dass trotzdem eine Fehlermeldung im Protokoll erscheint ... wie ist es denn die Prüfung auf swap vor die REAL funktione zu machen? Und diese dann erst gar nicht auszuführen? Man müsste aber noch nachschauen ob es für AVM okay ist die REAL nicht auszuführen.
Was natürlich auch noch fehlt ist das unmounten, es sollte auch machbar sein.
Allerdings finde ich swap-Dateien wesentlich eleganter. Diese kann man einfach in autorun/autoend einhängen
 D

**Comment by asmcc** [comment]
Wenn man es "umdreht", wird er auf SWAP IMMER checken, also wirklich bei jeder Partition. Genau das wolte ich vermeiden und es mehr idiotensicher machen, dass diese Änderungen nur wirklich ausgeführt werden, wenn es nötig ist. Warum im "Fehlerpfad"? Ganz einfach, von der Logik her: Für AVM ist eine SWAP-Partition eine "fehlerhafte" oder "unbekannte" Partition, daher ist es genau richtig, diese zusätzliche Prüfung dahin zu packen, wenn die Erkennung von AVM fehl schlägt. Dann erfolgt diese Zusatzprüfung nur wenn SWAP-Option aktiviert ist und eine Partition und nicht eine Datei da als SWAP unter FREETZ hinterlegt ist.
Der Fehler im AVM-Log ist wirklich nur "kosmetisch". Um den auszublenden, muss man in dem Teil der udev-Skripte von AVM "rumwildern", wie ich es früher bei FREETZMOUNT gemacht hatte und nicht in diesem Spezialteil, was FREETZ-spezifisch "ausgelagert" ist. Das Gleiche betrifft auch SWAP umount.
Ja, ich hatte bei FREETZMOUNT es damals mit SWAP sowohl für mount als auch für umount mit allen möglichen Verzweigungen und separaten Log-Meldungen gelöst, will es aber hier bei dieser udev-Variante möglichst einfach halten. Unter dem Motto "keep it simple" (nach KISS-Prinzip).
Die Bedenken bzgl. eines unsauberes umounts der SWAP-Partition stimmen nur bedingt. Denn wenn die box rebootet, wird am Ende "rc.swap stop" ausgeführt, was genau zu einem umount dieser SWAP-Partition führt. Es könnte höchstens zu Problemen kommen, wenn jemand versucht das komplete USB-Device über AVM-WebIF zu "entfernen" oder gar den USB-Stick einfach herausziehen. Aber dann ist einem solchen User sowieso nicht mehr zu helfen.
Ja, die Benutzung einer ganzen SWAP-Partition ist in den letzten Jahren ziemlich selten geworden, weil die heutige Generation der User sich vermutlich über sowas gar keine Gedanken macht. Dennoch würde mit meinem Patch es wenigstens korrekt behandelt und nicht wie jetzt fehlschlagen.
Wie gesagt, es ist meiner Meinung nach ziemlich einfach und idiotensicher aufgebaut und läuft auf meiner 7590 schon seit mehreren Tagen erfolgreich. Ich würde es einfach einchecken. Ist aber natürlich deine Entscheidung.

**Comment by fda77** [comment]
Okay, da es selten swap-partitionen sind sollte man es so prüfen wie du es gemacht hast. Soll die avm-Fehlermeldung halt bleiben wie sie ist

```
				[ -e /etc/init.d/rc.swap ] && /etc/init.d/rc.swap autostart "$_BD"
				_RV=$?
```
in _RV kann auch das Ergebnis der [] landen falls diese false ist, also wenn es rc.swap nicht gibt ist _RC=1

Ja, das unmounte übers avm-webif (das kann evtl auch mit ctlmgt_ctl anstossen) meinte ich. Wenn man schon swap einbaut könnte man es doch gleich komplett machen...
Sonst könnte man gleich alles via autorun/autoend machen

Ich hab auf meinem router noch 350 mb ram frei, selbst mit Cache sind 200 mb unbenutzt. Der ist wohl etwas überdimensioniert und swap liegt in weiter ferne

**Comment by fda77** [comment]
Am besten machst du das `if [ -e /etc/init.d/rc.swap ]` als äussere Bedingung, die Datei gibt es nur wenn man swap explizit aktiviert hat: https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/mod/mod.mk#L11

Die Ausführung von "modconf" ist teurer

**Comment by asmcc** [comment]
ok, mit dieser verkürzten if-Form hast du Recht. So rum habe ich gar nicht gedacht und einen Testfall ohne rc.swap hatte ich natürlich nicht.
`if [ -e /etc/init.d/rc.swap ]; then`
`  /etc/init.d/rc.swap autostart "$_BD"`
`  _RV=$?`
`fi`
wäre richtiger, ohne es jetzt geprüft zu haben. Ob man daraus einen Einzeiler erzeugen kann, wäre ich überfragt.

**Reply by fda77** [reply]
Vor "fi" und "then" noch je ein Semikolon
und vor _RV

**Reply by asmcc** [reply]
Ja, ich komme hier mit diesem komischen code-Modus durcheinander. Daher kommt es. Er geht hier mit enter innerhalb code-Tags anders um, als in IPPF

**Comment by fda77** [comment, marked as answer]
So vielleicht:
```
# Special case SWAP partition
if [ -e /etc/init.d/rc.swap ]; then
	if [ "$(modconf value MOD_SWAP mod)" == "yes" ]; then
		if [ "$(modconf value MOD_SWAP_FILE mod | grep '/dev')" == "$_BD" ]; then
			/etc/init.d/rc.swap autostart "$_BD"
			_RV=$?
		fi
	fi
fi
if [ "$_RV" == "0" ]; then
	eventadd 140 "SWAP Partition $_BD"
	udev_log $_RV "$_BD - mounted as SWAP partition"
else
	eventadd 142 "Partition $_BD" "$_FS"
	udev_log $_RV "$_BD - unsupported filesystem or partition table${_FS:+: $_FS}"
fi
```


Oder das gleiche als Einzeiler - zur besseren Lesbarkeit in mehrere Zeilen:
```
[ -e /etc/init.d/rc.swap ] && \
  [ "$(modconf value MOD_SWAP mod)" == "yes" ] && \
  [ "$(modconf value MOD_SWAP_FILE mod | grep '/dev')" == "$_BD" ] && \
  /etc/init.d/rc.swap autostart "$_BD" && \
  _RV=0
if [ "$_RV" == "0" ]; then
	eventadd 140 "SWAP Partition $_BD"
	udev_log $_RV "$_BD - mounted as SWAP partition"
else
	eventadd 142 "Partition $_BD" "$_FS"
	udev_log $_RV "$_BD - unsupported filesystem or partition table${_FS:+: $_FS}"
fi
```

**Comment by asmcc** [comment]
Lass uns mal bei der oberen Variante belassen. Sonst ist es mit diesen "line breaker" kein richtiger Einzeiler in dem Sinne.
Ich habe es jetzt in meinem geforkten repository upgedated. Zumindest für mich sieht es so aus, als ob es sich unter "pull request" auch upgedatet hat. Da ich allerdings nicht vom Fach bin und meine ersten Schritte mit GIT überhaupt gerade die Tage mache, prüf bitte, ob ich da alles richtig eingetragen habe.
Die zweite if-Schleife habe ich ebenfalls umgedreht und die numerische Prüfung durch symbolische ersetzt. Ist zwar alles Kleinigkeiten, aber so sieht es einheitlicher und besser "verdaubar" mit dem Rest des Codes.
Da wir es jetzt fünf mal umgedreht haben, würde ich es gerne noch auf meiner Box checken. Nicht, dass da jetzt doch noch was von der Logik her falsch ist.

**Reply by fda77** [reply]
Wenn du bei einem pull-request oben auf den "files changed" Reiter gehst siehst du was alles geändert wurde. Wenn man beim mergen "squash" nimmt wird der commit genau diese Änderung

**Comment by asmcc** [comment]
Also, geflasht und es läuft wie vorher.

![20230205_SWAP_mount_AVM](https://user-images.githubusercontent.com/40981206/216837901-3539858f-d0d6-481d-b02b-f06138645d4b.PNG)
![20230205_SWAP_mount_FREETZ](https://user-images.githubusercontent.com/40981206/216837910-06b0e6bb-4a28-4ce6-87c5-c817efe2affa.PNG)

Die Meldung im AVM-WebIF ist ja an sich nicht falsch und nicht widersprüchlich. Man muss es nur richtig interpretieren: Die spätere (aber höher stehende) Meldung im AVM-WebIF besagt, dass es in "unserem" libmodudevm.sh alles richtig durchlaufen ist. Die frühere (steht ganz unten) wird in diesem REAL-Modul von AVM erzeugt, also "außerhalb" von libmodudevm.sh. Also soll wohl heißen, dass es mit AVM-Mitteln nicht möglich war, die Partition einzubinden.

Wenn wir es andersrum gemacht hätten und AVM-Teil umgangen, wäre es meiner Meinung nach mehr "Seiteneffekte" zu erwarten, wie du es ja auch zurecht gemerkt hattest.

Von daher: Von mir aus kann eingecheckt werden.

**Comment by fda77** [comment]
Also ich fänd es noch gut wenn beim unmounten (übers avm-webif) auch irgend was in diese Richtugn wäre. Aber das ist deine Sache, ich mache in diese Richtung nichts

**Comment by asmcc** [comment]
Danke fürs Einchecken!
Lass uns es so belassen. Sollten in dem Zusammenhang Nebenwirkungen entstehen (was ich nicht glaube) kann man es sich nochmal anschauen.

**Comment by fda77** [comment]
So ganz zufrieden bin nicht. In udevmount befindet sich nun swap-logik. Diese 2 if:
https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/mod/files/root/usr/lib/libmodudevm.sh#L27

Das ganze ist doppelt da es sich auch in der rc.swap befinden:
https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/mod/files/root/etc/init.d/rc.swap#L52

Aber vorerst soll es so okay sein
